Frontend/datepicker enhancements

rmolinares opened the pull request:
### Requirements List
- _None_

### Description List
- Added default start date for registration DOB calendar to be 1975
- Added ability to use `InputDate` component with option to only trigger picker by clicking on calender icon and allow input typing without triggering the picker
- Added methods to auto-format date field if typed in manually
- Should work cross device

### Testing List
- `yarn test:unit:all` should run without errors or warnings
- `yarn serve` should run without errors or warnings
- `yarn build` should run without errors or warnings
- For API configuration changes: CDK tests added/updated in `backend/compact-connect/tests/unit/test_api.py`
- Code review
- Testing
    - On the practitioner registration screen, confirm that for DOB you can click on the input (_not_ the calendar icon) and manually type in a date that is auto formatted as you type (MM/dd/yyyy)
    - Confirm that you are also able to click the related calendar icon and that is what triggers the datepicker UI
    - Confirm that the default year in the UI is 1975
    - Confirm functionality works in desktop as well as mobile identically
    - Confirm tabbing works and lets you proceed as expected

Closes #826 


<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

- **New Features**
	- Enhanced date input fields with improved parsing, formatting, and validation for user-entered dates.
	- Added support for a custom text input mode and formatting options in date pickers.
	- Introduced a custom input UI for date selection when the calendar menu is disabled.
- **Style**
	- Improved accessibility and keyboard navigation for date input fields.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

rmolinares commented on webroot/src/components/Forms/InputDate/InputDate.ts:
```diff
@@ -56,6 +62,93 @@
+            const dayNum = parseInt(day, 10);
+            const yearNum = parseInt(year, 10);
+
+            if (monthNum >= 1 && monthNum <= 12 && dayNum >= 1 && dayNum <= 31 && yearNum >= 1900 && yearNum <= 2100) {
+                raw = `${year}-${month}-${day}`;
+            }
```
Your suggest change triggers a linter error:
`Object is possibly 'undefined'` referring to this.yearRange[0] and this.yearRange[1]

rmolinares commented:
@jlkravitz This is ready for your review.

jlkravitz commented on webroot/src/components/Forms/InputDate/InputDate.ts:
```diff
@@ -56,9 +70,77 @@ class InputDate extends mixins(MixinInput) {
+        return dateInputToServerFormat(this.localValue.replace(/\D/g, ''));
+    }
+
+    get isDateValid(): boolean {
```
I don't see this function being used anywhere, so perhaps we can remove.

jlkravitz replied:
And as a result, also true of `isValidDateInput`?

jlkravitz commented on webroot/src/components/Forms/InputDate/InputDate.ts:
```diff
@@ -56,9 +70,77 @@ class InputDate extends mixins(MixinInput) {
+    //
+    // Computed
+    //
+    get dateDisplay(): string {
+        return formatDateInput(this.localValue);
+    }
+
+    set dateDisplay(value: string) {
```
Are these used anywhere? If not, can we remove?

rmolinares replied:
Good catch. Removed.

jlkravitz reviewed (commented):
few small things!

rmolinares commented on webroot/src/components/Forms/InputDate/InputDate.ts:
```diff
@@ -56,9 +70,77 @@ class InputDate extends mixins(MixinInput) {
     //
+    onInput(): void {
+        // Format the input as user types
+        const numericOnly = this.localValue.replace(/\D/g, '');
```
Yeah. This is redundant. Removed.

rmolinares commented:
@jlkravitz I've addressed your initial feedback. I also noticed some minor edge cases in the UI so I refactored a little on my latest commit as well. Feel free to give it another review when you can. Thanks.

jlkravitz commented on webroot/src/models/_formatters/date.ts:
```diff
@@ -101,9 +101,79 @@ const dateDiff = (date1: stringOptional, date2: stringOptional, diffUnit: unitOf
+    return formatted;
+};
+
+/**
+ * Convert numeric date input (MMddyyyy) to server format (yyyy-MM-dd)
+ * @param {string} value Raw numeric input (e.g., "12252024")
+ * @return {string} Server format date string (e.g., "2024-12-25") or empty string if invalid
+ */
+const dateInputToServerFormat = (value: string): string => {
+    const numericOnly = (value || '').replace(/\D/g, '');
```
Can we update the docs here now since technically you can have non-digit characters which are stripped?

jlkravitz commented on webroot/src/models/_formatters/date.ts:
```diff
@@ -101,9 +101,79 @@ const dateDiff = (date1: stringOptional, date2: stringOptional, diffUnit: unitOf
 };
 
+/**
+ * Format numeric date input (MMddyyyy) to MM/dd/yyyy display format
```
Can we update the docs here now since technically you can have non-digit characters which are stripped?

jlkravitz commented on webroot/src/components/Forms/InputDate/InputDate.ts:
```diff
@@ -56,9 +70,64 @@ class InputDate extends mixins(MixinInput) {
+    // Computed
+    //
+    get dateRaw(): string {
+        return dateInputToServerFormat(this.localValue.replace(/\D/g, ''));
```
can't we remove the `replace` call?

jlkravitz reviewed (commented):
looks good! two very small things

rmolinares commented:
> looks good! two very small things

@jlkravitz Yes, good catches on those! I've now resolved them and ready for another review.

jlkravitz reviewed (approved):
@isabeleliassen ready to merge!
